### PR TITLE
Chore: Define promlib depguard rules

### PR DESCRIPTION
**Pull request by itsmylife**
**What is this feature?**

Define rules to prevent accidental `grafana/grafana` imports in promlib.
https://github.com/OpenPeeDeeP/depguard?tab=readme-ov-file#golangci-lint

After creating [a commit](https://github.com/grafana/grafana/pull/85082/commits/9f0538859744bec798e531e4de02140d3f13c151) that has a code against rules the linter gives us the error, [see here](https://drone.grafana.net/grafana/grafana/169001/4/6).

Successful pipeline: https://drone.grafana.net/grafana/grafana/169004/4/6 

**Who is this feature for?**

promlib developers

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: .golangci.toml [hunk 1]
```diff
@@ -67,6 +67,19 @@ files = [
   "**/pkg/tsdb/cloudwatch/**/*",
 ]
 
+[linters-settings.depguard.rules.promlib]
+list-mode = "lax" # allow unless explicitely denied
+deny = [
+  { pkg = "github.com/grafana/grafana/pkg", desc = "promlib is not allowed to import grafana core" }
```
Comment (toddtreece) [inline .golangci.toml:73]: what happens if this is `github.com/grafana/grafana` instead of `github.com/grafana/grafana/pkg`?

**Review by toddtreece — APPROVED** [review]

File: .golangci.toml (hunk 1 above)
Comment (itsmylife) [reply .golangci.toml:73]: Oh 🤦 that should be exactly the way you said. I forgot that while I was testing stuff out. 
Nice catch!

Comment (itsmylife) [reply .golangci.toml:73]: If I put `github.com/grafana/grafana` I get a ton of errors:

<details>
<summary>Expand for Errors</summary>

```
❯ make golangci-lint
lint via golangci-lint
/Users/ismailsimsek/go/bin/golangci-lint-v1.57.1 run \
                --config .golangci.toml \
                ./pkg/... ./pkg/apiserver/... ./pkg/apimachinery/... ./pkg/promlib/...
pkg/promlib/models/query.go:10:2: import 'github.com/grafana/grafana-plugin-sdk-go/backend' is not allowed from list 'promlib': promlib is not allowed to import grafana core (depguard)
        "github.com/grafana/grafana-plugin-sdk-go/backend"
        ^
pkg/promlib/models/query.go:11:2: import 'github.com/grafana/grafana-plugin-sdk-go/backend/gtime' is not allowed from list 'promlib': promlib is not allowed to import grafana core (depguard)
        "github.com/grafana/grafana-plugin-sdk-go/backend/gtime"
        ^
pkg/promlib/models/result.go:6:2: import 'github.com/grafana/grafana-plugin-sdk-go/data' is not allowed from list 'promlib': promlib is not allowed to import grafana core (depguard)
        "github.com/grafana/grafana-plugin-sdk-go/data"
        ^
pkg/promlib/utils/utils.go:8:2: import 'github.com/grafana/grafana-plugin-sdk-go/backend' is not allowed from list 'promlib': promlib is not allowed to import grafana core (depguard)
        "github.com/grafana/grafana-plugin-sdk-go/backend"
        ^
pkg/promlib/querydata/exemplar/framer.go:6:2: import 'github.com/grafana/grafana-plugin-sdk-go/data' is not allowed from list 'promlib': promlib is not allowed to import grafana core (depguard)
        "github.com/grafana/grafana-plugin-sdk-go/data"
        ^
pkg/promlib/querydata/exemplar/labels.go:6:2: import 'github.com/grafana/grafana-plugin-sdk-go/data' is not allowed from list 'promlib': promlib is not allowed to import grafana core (depguard)
        "github.com/grafana/grafana-plugin-sdk-go/data"
        ^
pkg/promlib/resource/resource.go:9:2: import 'github.com/grafana/grafana-plugin-sdk-go/backend' is not allowed from list 'promlib': promlib is not allowed to import grafana core (depguard)
        "github.com/grafana/grafana-plugin-sdk-go/backend"
        ^
pkg/promlib/resource/resource.go:10:2: import 'github.com/grafana/grafana-plugin-sdk-go/backend/log' is not allowed from list 'promlib': promlib is not allowed to import grafana core (depguard)
        "github.com/grafana/grafana-plugin-sdk-go/backend/log"
        ^
pkg/promlib/resource/resource.go:11:2: import 'github.com/grafana/grafana-plugin-sdk-go/data/utils/maputil' is not allowed from list 'promlib': promlib is not allowed to import grafana core (depguard)
        "github.com/grafana/grafana-plugin-sdk-go/data/utils/maputil"
        ^
pkg/promlib/heuristics.go:11:2: import 'github.com/grafana/grafana-plugin-sdk-go/backend/log' is not allowed from list 'promlib': promlib is not allowed to import grafana core (depguard)
        "github.com/grafana/grafana-plugin-sdk-go/backend/log"
        ^
pkg/promlib/library.go:11:2: import 'github.com/grafana/grafana-plugin-sdk-go/backend/datasource' is not allowed from list 'promlib': promlib is not allowed to import grafana core (depguard)
        "github.com/grafana/grafana-plugin-sdk-go/backend/datasource"
        ^
pkg/promlib/library.go:12:2: import 'github.com/grafana/grafana-plugin-sdk-go/backend/httpclient' is not allowed from list 'promlib': promlib is not allowed to import grafana core (depguard)
        sdkhttpclient "github.com/grafana/grafana-plugin-sdk-go/backend/httpclient"
        ^
pkg/promlib/library.go:13:2: import 'github.com/grafana/grafana-plugin-sdk-go/backend/instancemgmt' is not allowed from list 'promlib': promlib is not allowed to import grafana core (depguard)
        "github.com/grafana/grafana-plugin-sdk-go/backend/instancemgmt"
        ^
pkg/promlib/library.go:14:2: import 'github.com/grafana/grafana-plugin-sdk-go/backend/log' is not allowed from list 'promlib': promlib is not allowed to import grafana core (depguard)
        "github.com/grafana/grafana-plugin-sdk-go/backend/log"
        ^
pkg/promlib/healthcheck_test.go:12:2: import 'github.com/grafana/grafana-plugin-sdk-go/backend/datasource' is not allowed from list 'promlib': promlib is not allowed to import grafana core (depguard)
        "github.com/grafana/grafana-plugin-sdk-go/backend/datasource"
        ^
pkg/promlib/healthcheck_test.go:13:2: import 'github.com/grafana/grafana-plugin-sdk-go/backend/httpclient' is not allowed from list 'promlib': promlib is not allowed to import grafana core (depguard)
        sdkhttpclient "github.com/grafana/grafana-plugin-sdk-go/backend/httpclient"
        ^
pkg/promlib/heuristics_test.go:15:2: import 'github.com/grafana/grafana-plugin-sdk-go/backend/datasource' is not allowed from list 'promlib': promlib is not allowed to import grafana core (depguard)
        "github.com/grafana/grafana-plugin-sdk-go/backend/datasource"
        ^
pkg/promlib/heuristics_test.go:16:2: import 'github.com/grafana/grafana-plugin-sdk-go/backend/httpclient' is not allowed from list 'promlib': promlib is not allowed to import grafana core (depguard)
        sdkhttpclient "github.com/grafana/grafana-plugin-sdk-go/backend/httpclient"
        ^
pkg/promlib/client/transport.go:10:2: import 'github.com/grafana/grafana-plugin-sdk-go/data/utils/maputil' is not allowed from list 'promlib': promlib is not allowed to import grafana core (depguard)
        "github.com/grafana/grafana-plugin-sdk-go/data/utils/maputil"
        ^
pkg/promlib/converter/prom.go:11:2: import 'github.com/grafana/grafana-plugin-sdk-go/data/utils/jsoniter' is not allowed from list 'promlib': promlib is not allowed to import grafana core (depguard)
        sdkjsoniter "github.com/grafana/grafana-plugin-sdk-go/data/utils/jsoniter"
        ^
pkg/promlib/converter/prom_test.go:10:2: import 'github.com/grafana/grafana-plugin-sdk-go/data/utils/jsoniter' is not allowed from list 'promlib': promlib is not allowed to import grafana core (depguard)
        sdkjsoniter "github.com/grafana/grafana-plugin-sdk-go/data/utils/jsoniter"
        ^
pkg/promlib/converter/prom_test.go:11:2: import 'github.com/grafana/grafana-plugin-sdk-go/experimental' is not allowed from list 'promlib': promlib is not allowed to import grafana core (depguard)
        "github.com/grafana/grafana-plugin-sdk-go/experimental"
        ^
pkg/promlib/intervalv2/intervalv2.go:8:2: import 'github.com/grafana/grafana-plugin-sdk-go/backend/gtime' is not allowed from list 'promlib': promlib is not allowed to import grafana core (depguard)
        "github.com/grafana/grafana-plugin-sdk-go/backend/gtime"
        ^
pkg/promlib/querydata/request.go:12:2: import 'github.com/grafana/grafana-plugin-sdk-go/data/utils/maputil' is not allowed from list 'promlib': promlib is not allowed to import grafana core (depguard)
        "github.com/grafana/grafana-plugin-sdk-go/data/utils/maputil"
        ^
pkg/promlib/querydata/request.go:17:2: import 'github.com/grafana/grafana-plugin-sdk-go/backend/tracing' is not allowed from list 'promlib': promlib is not allowed to import grafana core (depguard)
        "github.com/grafana/grafana-plugin-sdk-go/backend/tracing"
        ^
pkg/promlib/querydata/framing_test.go:16:2: import 'github.com/grafana/grafana-plugin-sdk-go/experimental' is not allowed from list 'promlib': promlib is not allowed to import grafana core (depguard)
        "github.com/grafana/grafana-plugin-sdk-go/experimental"
        ^
pkg/promlib/querydata/exemplar/sampler_stddev_test.go:7:2: import 'github.com/grafana/grafana-plugin-sdk-go/experimental' is not allowed from list 'promlib': promlib is not allowed to import grafana core (depguard)
        "github.com/grafana/grafana-plugin-sdk-go/experimental"
        ^
make: *** [golangci-lint] Error 1

```

</details>

Comment (toddtreece) [reply .golangci.toml:73]: 👍 i just tested that in #84084 and hit the same issue

Comment (itsmylife) [reply .golangci.toml:73]: @toddtreece I am merging this if that's ok then?

**Pull request merged**